### PR TITLE
Macro indirection for stepper drivers + Support for TMC26X and L6470 Stepper Drivers

### PR DESCRIPTION
**Pull request by domonoky**
This patch introduces a layer of macro indirection to allow support for other stepper drivers which are not entirely controlled by i/o pins. 
The macro layer makes sure there is no performance loss for normal I/O pin driven stepper drivers.

This patch also adds support for stepper drivers based on TMC26 chips and L6470 chips using arduino libraries.

I am happy to hear you comments or suggestions for improvements.

### CODE REVIEW

**Comment by thinkyhead** [comment]
Looks good! But it appears you'll need to modify the `.travis.yml` file to get this to include the new libraries and get through a verification build (via Travis CI). And/or, you can wrap all the new #include statements (for example, `#include <TMC26XStepper.h>`) in conditionals so that the build system won't try to include them and fail.

**Comment by domonoky** [comment]
I am not sure why it even looks for this header. All those includes are wrapped in #ifdefs for the features and those ifdefs should not be enabled. 
I would be happy if someone could point me in the right direction to fix this..

**Comment by nophead** [comment]
Travis build use ino, which does not understand conditional includes.

**Comment by thinkyhead** [comment]
@domonoky Yes, I just had a closer look and it's strange that it's looking for that header. I'm re-running the test, which is rebooting the vm, but results will probably be the same. Maybe it's because there are other .ino files in the bundle. It may be trying to compile the library's included test. But, hmm... probably not that either. Messing with the .yml file might cause it behave differently.

**Comment by domonoky** [comment]
I changed the travis.yml file to copy the required libraries into the arduino library folder, and now travis works.
I also checked (with Arduino 1.5.8) if it can be built locally without the libraries installed, and it seems to work. Would be good if someone else can double-check that. (maybe with an older arduino version)

**Comment by thinkyhead** [comment]
@domonoky Aha. I _almost_ forgot 1.5.8 can do that. If we can require Arduino 1.5.8 or later I have a PR to move some of Marlin into subfolders, by making parts of it into local "libraries."

File: Marlin/stepper.cpp [hunk 1]
```diff
@@ -959,51 +969,51 @@ void st_init()
 
   //Initialize Step Pins
   #if defined(X_STEP_PIN) && (X_STEP_PIN > -1)
-    SET_OUTPUT(X_STEP_PIN);
-    WRITE(X_STEP_PIN,INVERT_X_STEP_PIN);
+    X_STEP_INIT;
+    X_STEP_WRITE(INVERT_X_STEP_PIN);
     disable_x();
```
Comment (thinkyhead) [inline Marlin/stepper.cpp]: A new macro called `OUT_WRITE(p,v)` combines `SET_OUTPUT(p)` with `WRITE(p,v)`. If you do an upstream merge from Development you might get a conflict on these lines. When you resolve the conflict, your new code here should supersede the lines that have `OUT_WRITE` in them now. If you feel like following the same single-macro pattern you could add macros like...

    #define X_STEP_INIT_WRITE(v) OUT_WRITE(X_STEP_PIN,v)

File: Marlin/stepper.cpp [hunk 2]
```diff
@@ -959,51 +969,51 @@ void st_init()
 
   //Initialize Step Pins
   #if defined(X_STEP_PIN) && (X_STEP_PIN > -1)
-    SET_OUTPUT(X_STEP_PIN);
-    WRITE(X_STEP_PIN,INVERT_X_STEP_PIN);
+    X_STEP_INIT;
+    X_STEP_WRITE(INVERT_X_STEP_PIN);
```
Comment (thinkyhead) [inline Marlin/stepper.cpp:966]: A new macro called `OUT_WRITE(p,v)` combines `SET_OUTPUT(p)` with `WRITE(p,v)`. If you do an upstream merge from Development you might get a conflict on these lines. When you resolve the conflict, your new code here should supersede the lines that have `OUT_WRITE` in them now. If you feel like following the same single-macro pattern you could add macros like...

```
#define X_STEP_INIT_WRITE(STATE) OUT_WRITE(X_STEP_PIN,STATE)
```

Comment (domonoky) [inline Marlin/stepper.cpp:966]: Thank you for the heads up. I will do that in the next days and bring back in sync.

**Comment by domonoky** [comment]
The merge conflicts should now be resolved, and everythign be working again.

**Comment by thinkyhead** [comment]
Looks good!

**Pull request merged**